### PR TITLE
Improve CodeAction to insert expected elements to support choice

### LINKED ISSUE #1448: Improve CodeAction "Insert only required expected elements" by taking care of choice
Given this XML file:

```xml
<?xml version="1.0" encoding="UTF-8"?>
<?xml-model href="http://www.tei-c.org/release/xml/tei/custom/schema/relaxng/tei_all.rng" type="application/xml" schematypens="http://relaxng.org/ns/structure/1.0"?>
<TEI xmlns="http://www.tei-c.org/ns/1.0">
	<teiHeader>
		<fileDesc>
			<titleStmt>
				<title></title>
			</titleStmt>
			<publicationStmt>
				<publisher></publisher>
			</publicationStmt>
			<sourceDesc>
				<p></p>
			</sourceDesc>
		</fileDesc>
	</teiHeader>
	<text>

	</text>
</TEI>
```

As `text` element can hosts several element kind (addSpan, etc), it should be nice to provide a code action per possible element to generate the content of text.

image

### PR DESCRIPTION
**Pull request by JessicaJHee**
Fixes #1448 

Signed-off-by: Jessica He <jhe@redhat.com>

### CODE REVIEW

**Comment by JessicaJHee** [comment]
Given the rng schema:

**Comment by angelozerr** [comment]
The PR looks very good.

I think it should be nice to support resolve code action to avoid computing textedit on the hover but when code action is clicked (for this code action and for the last that you implemented). You can do  that in a separate PR if you wish.

**Comment by JessicaJHee** [comment]
> I think it should be nice to support resolve code action to avoid computing textedit on the hover but when code action is clicked (for this code action and for the last that you implemented). You can do that in a separate PR if you wish.

I will try to do that in this PR!

**Comment by angelozerr** [comment]
@JessicaJHee I suggest that you play with pretext rng (you can include the pretext.rng in your test resources) and you will see some trouble.

Given this XML file:

```xml
<?xml-model href="https://raw.githubusercontent.com/PreTeXtBook/pretext/master/schema/pretext.rng"?>
<pretext>
	
</pretext>
```

If you try to process `book` code action, it generates:

```xml
<?xml-model href="https://raw.githubusercontent.com/PreTeXtBook/pretext/master/schema/pretext.rng"?>
<pretext>
	<book>
		<title></title>
		<part>
			<chapter>
				<p></p>
			</chapter>
		</part>
	</book>
</pretext>
```

It misses the title element for part and chapter:

image

In otherwise it should generate:

```xml
<?xml-model href="https://raw.githubusercontent.com/PreTeXtBook/pretext/master/schema/pretext.rng"?>
<pretext>
	<book>
		<title></title>
		<part>
			<title></title>
			<chapter>
				<title></title>
				<p></p>
			</chapter>
		</part>
	</book>
</pretext>
```

I suspect that it is because we want to avoid recusrsion and we don't generate title for part and chapter although we need to generate it. I think we should generate title if is is required. But it requires some more investigation.

**Comment by JessicaJHee** [comment]
This PR should also fix https://github.com/eclipse/lemminx/issues/1278, since a better method is used to keep track of generated elements.

**Comment by datho7561** [comment]
I tried out the schema you posted with the following document:

```xml
<?xml-model href="choiseSchema.rng"?>
<article></article>
```

When I select the "Insert expected element 'title'" code action, the opening tag is generated, but the closing tag isn't. For the other two code actions, something similar happens (opening tag is generated but closing isn't). Is this expected?

**Comment by JessicaJHee** [comment]
> When I select the "Insert expected element 'title'" code action, the opening tag is generated, but the closing tag isn't. For the other two code actions, something similar happens (opening tag is generated but closing isn't). Is this expected?

Hmm no that's weird, on my end it works the same as the gif I posted earlier. I'm not sure why that happens, do you have any non-default settings set by any chance?

**Comment by datho7561** [comment]
I disabled auto close tags. Re-enabling auto close tags seems to make the code action behave as expected.

**Review by datho7561 — CHANGES_REQUESTED** [review]
Sorry it took so long, I ran into some bugs unrelated to the PR that I spent time to file.

Two more things to consider:
- XSD schemas also have a `choice` mechanism. I think it might be helpful to provide a similar code action for XSD at some point, although we should see if people want that first, and if so, do that in a different PR
- If the child of the element that you pick is a choice between several elements, then it will automatically generate the first valid choice. In the future, it might be helpful to stop generating children of children when you reach a choice.

For example, given the following schema:

```xml
<grammar xmlns="http://relaxng.org/ns/structure/1.0"
  datatypeLibrary="http://www.w3.org/2001/XMLSchema-datatypes">
  <start>
    <ref name="root" />
  </start>
  <define name="root">
    <element name="root">
      <choice>
        <element name="child1">
          <choice>
            <element name="child1A">
              <text />
            </element>
            <element name="child1B">
              <text />
            </element>
          </choice>
        </element>
        <element name="child2">
          <choice>
            <element name="child2A">
              <text />
            </element>
            <element name="child2B">
              <text />
            </element>
          </choice>
        </element>
      </choice>
    </element>
  </define>
</grammar>
```

If you have the XML document `<root></root>` and you generate `child1`, currently it generates `child1A` as well. I think it might be better to only generate `child1` so that the user has the option to pick between `child1A` and `child1B`. We can do this in a future PR, though

**Comment by angelozerr** [comment]
> If you have the XML document <root></root> and you generate child1, currently it generates child1A as well. I think it might be better to only generate child1 so that the user has the option to pick between child1A and child1B. We can do this in a future PR, though

David I understand your point of view, but my idea was to provide:

 * a code action which fix a problem and doesn't report a new error.
 * a code action which fill quickly an expected XML content.

I think your code action idea can be interesting too, but for me it is a new code action. We need just to find a proper title.

**Comment by datho7561** [comment]
> I think your code action idea can be interesting too, but for me it is a new code action. We need just to find a proper title.

Okay. Sounds good. I'll file an issue for this.

File: org.eclipse.lemminx/src/main/java/org/eclipse/lemminx/extensions/contentmodel/participants/codeactions/missingelement/required_element_missingCodeActionResolver.java [hunk 1]
```diff
@@ -10,77 +10,71 @@
...
-
-			CodeAction insertAllExpectedElement = CodeActionFactory.replace("Insert all expected elements",
-					targetRange, insertStrAll, document.getTextDocument(), diagnostic);
-
-			codeActions.add(insertAllExpectedElement);
-
-		} catch (
-
-		BadLocationException e) {
+			if ("required".equals(request.getDataProperty(DATA_REQUIRED_FIELD))) {
```
[the remark this answers is not in the document]
Comment (JessicaJHee) [reply org.eclipse.lemminx/src/main/java/org/eclipse/lemminx/extensions/contentmodel/participants/codeactions/missingelement/required_element_missingCodeActionResolver.java]: Sounds good! I've changed it to "onlyGenerateRequired"

**Comment by angelozerr** [comment]
@JessicaJHee when there are a lot of elements choice, it can be hard to select the proper code action. I suggest that you sort the code action by element name.

 * Insert expected element `anchor`
 * Insert expected element `body`
 * ...

instead of displaying:

 * Insert expected element `body`
 * Insert expected element `anchor`
 * ...

File: org.eclipse.lemminx/src/main/java/org/eclipse/lemminx/extensions/contentmodel/participants/codeactions/missingelement/required_elements_missing_expectedCodeActionResolver.java [hunk 2]
```diff
@@ -0,0 +1,89 @@
...
+			Range targetRange = new Range(childElementPositionStartTag, childElementPositionEndTag);
+			XMLGenerator generator = request.getXMLGenerator(targetRange);
+
+			ContentModelManager contentModelManager = request.getComponent(ContentModelManager.class);
+
+			for (CMDocument cmDocument : contentModelManager.findCMDocument(element)) {
+				CMElementDeclaration matchesCMElement = cmDocument.findCMElement(element);
+				if (matchesCMElement != null) {
+					// Add a code action option for each element choice
+					for (CMElementDeclaration child : matchesCMElement.getElements()) {
```
Comment (angelozerr) [inline org.eclipse.lemminx/src/main/java/org/eclipse/lemminx/extensions/contentmodel/participants/codeactions/missingelement/required_elements_missing_expectedCodeActionResolver.java:72]: I will introduce a new method in CMElementDeclaration:

`for (CMElementDeclaration child : matchesCMElement.getPossibleRequiredElements()) {`

and create class :

```java
/*******************************************************************************
* Copyright (c) 2023 Red Hat Inc. and others.
* All rights reserved. This program and the accompanying materials
* which accompanies this distribution, and is available at
* http://www.eclipse.org/legal/epl-v20.html
*
* SPDX-License-Identifier: EPL-2.0
*
* Contributors:
*     Red Hat Inc. - initial API and implementation
*******************************************************************************/
package com.thaiopensource.relaxng.pattern;

import java.util.Set;

public class PossibleRequiredElementsFunction extends AbstractPatternFunction<Void> {

	private final Set<String> requiredElementNames;

	public PossibleRequiredElementsFunction(Set<String> requiredElementNames) {
		this.requiredElementNames = requiredElementNames;
	}

	@Override
	public Void caseOther(Pattern p) {
		return null;
	}

	@Override
	public Void caseElement(ElementPattern p) {
		return caseNamed(p.getNameClass());
	}

	@Override
	public Void caseGroup(GroupPattern p) {
		return union(p);
	}

	@Override
	public Void caseChoice(ChoicePattern p) {
		// For choice, get the first pattern which is not nullable
//		Pattern p1 = p.getOperand1();
//		if (!p1.isNullable()) {
//			p1.apply(this);
//		} else {
//			Pattern p2 = p.getOperand2();
//			if (!p2.isNullable()) {
//				p2.apply(this);
//			}
//		}
//		return null;
		return union(p);
	}

	@Override
	public Void caseInterleave(InterleavePattern p) {
		return union(p);
	}

	@Override
	public Void caseAfter(AfterPattern p) {
		return p.getOperand1().apply(this);
	}

	@Override
	public Void caseOneOrMore(OneOrMorePattern p) {
		return p.getOperand().apply(this);
	}

	private Void caseNamed(NameClass nc) {
		if (!(nc instanceof SimpleNameClass))
			return null;
		requiredElementNames.add(((SimpleNameClass) nc).getName().getLocalName());
		return null;
	}
	
	private Void union(BinaryPattern p) {
		Pattern p1 = p.getOperand1();
		if (!p1.isNullable()) {
			p1.apply(this);
		}
		Pattern p2 = p.getOperand2();
		if (!p2.isNullable()) {
			p2.apply(this);
		}
		return null;
	}
}

```

**Comment by angelozerr** [comment]
Wow it is impressive, it works now like a charm. Thanks so much @JessicaJHee !

**Comment by angelozerr** [comment]
@oscarlevin here a demo with Pretext:

PretextCodeAction

**Pull request merged**